### PR TITLE
Dashboard

### PR DESCRIPTION
**Pull request by Pilou97**
# Context

The goal is to add a page `/dashboard` which will be the default page when using a wallet.

The dashboard many informations:
 - tez balance
 - Defi token balances
 - XTZ price in dollar
 - price evolution
 - token repartitions in a beautiful chart

This merge request solves the issue: https://github.com/marigold-dev/tzsafe-ui/issues/78

### CODE REVIEW

File: components/dashboard/assets.tsx [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import { defi } from "../../utils/tzktHooks";
+
+type AssetsProperty = {
```
Comment (quentin-burg) [inline components/dashboard/assets.tsx:3]: Généralement la convention de nommage pour les props c'est <nom>Props
Par exemple `AssetsProps`

File: components/dashboard/balance.tsx [hunk 2]
```diff
@@ -0,0 +1,31 @@
+type BalanceProperty = {
```
Comment (quentin-burg) [inline components/dashboard/balance.tsx:1]: Idem pour Props

File: components/dashboard/donut.tsx [hunk 3]
```diff
@@ -0,0 +1,86 @@
+import { Chart as ChartJS, ArcElement, Legend, Tooltip } from "chart.js";
+import { Doughnut } from "react-chartjs-2";
+import { defi } from "../../utils/tzktHooks";
+
+ChartJS.register(ArcElement, Tooltip, Legend);
+
+type DonutProperty = {
```
Comment (quentin-burg) [inline components/dashboard/donut.tsx:7]: Idem pour Props

File: components/dashboard/donut.tsx [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  title: string;
+  colors: string[];
+  tokens: defi[];
+};
+
+const Donut = ({ title, colors, tokens }: DonutProperty) => {
+  // We only need 4 tokens, the 5th one is the sum of the remaining
+  const tokenSample: Array<defi> = tokens.reduceRight<Array<defi>>(
```
Comment (quentin-burg) [inline components/dashboard/donut.tsx:15]: Peut etre utiliser `useCallback` pour éviter les recalculs inutiles, tu veux juste recalculer quand `tokens` change

File: components/dashboard/index.tsx [hunk 5]
```diff
@@ -0,0 +1,50 @@
+import { defi, price } from "../../utils/tzktHooks";
+import Assets from "./assets";
+import Balance from "./balance";
+import Donut from "./donut";
+
+type DashboardProperty = {
```
Comment (quentin-burg) [inline components/dashboard/index.tsx:6]: Idem pour Props

File: utils/tzktHooks.ts [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+      name: string | null;
+      symbol: string | null;
+      decimals: string | null;
+      thumbnailUri: string | null; // this is not documented in the Tzkt API
+    } | null;
+  };
+  balance: string;
+};
+
+export type defi = {
```
Comment (quentin-burg) [inline utils/tzktHooks.ts:50]: De préférence les types commencent avec une majuscule

File: utils/tzktHooks.ts [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+  return balance;
+};
+
+/**
+ * Token type returned by:
+ * https://api.tzkt.io/v1/tokens/balances
+ * Only useful fields are typed
+ */
+type tzkt_token = {
```
Comment (quentin-burg) [inline utils/tzktHooks.ts:32]: Type = MAJ :-p

File: utils/tzktHooks.ts [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+import { TZKT_API_URL } from "../context/config";
+
+/**
+ * Returns the balance of the given address
+ * @param address tz/kt Tezos address
+ */
+export const useTzktBalance = (address: string | null) => {
+  const [balance, setBalance] = useState(0);
+  useEffect(() => {
+    if (address == null) return;
```
Comment (quentin-burg) [inline utils/tzktHooks.ts:11]: `===`

File: utils/tzktHooks.ts [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+    (async () => {
+      const response = await fetch(
+        `${TZKT_API_URL}/v1/accounts/${address}/balance`
+      );
+
+      if (response.status == 200) {
```
Comment (quentin-burg) [inline utils/tzktHooks.ts:17]: `===`

File: utils/tzktHooks.ts [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+
+/**
+ * Returns Defi Tokens of the given address
+ * @param address tz/kt Tezos address
+ */
+export const useTzktDefiTokens = (address: string | null) => {
+  const [defi, setDefi] = useState<defi[]>([]);
+
+  useEffect(() => {
+    if (address == null) return;
```
Comment (quentin-burg) [inline utils/tzktHooks.ts:65]: `===`

File: utils/tzktHooks.ts [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+      // Check if the response is a success
+      if (response.status !== 200) {
+        return;
+      }
+      const tokens: Array<tzkt_token> = await response.json();
+      // First removes what is not a DeFi token
+      const defi = tokens
+        .filter(token => {
+          return (
+            token.token.metadata != null &&
```
Comment (quentin-burg) [inline utils/tzktHooks.ts:79]: `!==`

Comment (quentin-burg) [reply utils/tzktHooks.ts:79]: Enfin ça dépend ce que tu veux tester, c'est ptet `undefined` ?

File: utils/tzktHooks.ts [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+          return { contract, balance, symbol, icon };
+        })
+        .filter(token => token.balance !== 0); // We don't care about tokens with 0 balance, I prefer to filter here because the balance is converted to number
+      setDefi(defi);
+    })();
+  }, [address]);
+  return defi;
+};
+
+export type price = {
```
Comment (quentin-burg) [inline utils/tzktHooks.ts:109]: type = MAJ :-p

File: utils/tzktHooks.ts [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+ * */
+export const useTzktPrice = (currency = "usd") => {
+  const [price, setPrice] = useState<price | null>(null);
+  useEffect(() => {
+    (async () => {
+      // Let's compute the price evolution
+      const pricesResponse = await fetch(
+        `https://back.tzkt.io/v1/home?quote=${currency}`
+      );
+      if (pricesResponse.status != 200) return;
```
Comment (quentin-burg) [inline utils/tzktHooks.ts:127]: `!==`

File: utils/tzktHooks.ts [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+        const date2 = new Date(price2.date);
+        return date1.getTime() - date2.getTime();
+      });
+      const first = prices[0];
+      const last = prices[prices.length - 1];
+      const evolution = last.value / first.value;
+
+      // Let's get the current price
+      const headResponse = await fetch("https://back.tzkt.io/v1/head");
+      if (headResponse.status != 200) return;
```
Comment (quentin-burg) [inline utils/tzktHooks.ts:143]: `!==` et `===`

**Review by quentin-burg — APPROVED** [review]
Few things to change, but GG 💪

File: utils/tzktHooks.ts (hunk 14 above)
Comment (Pilou97) [reply utils/tzktHooks.ts]: fixed in: https://github.com/marigold-dev/tzsafe-ui/pull/167/commits/d6061725be57076a19783e218f046ffbbed80438

File: utils/tzktHooks.ts (hunk 13 above)
Comment (Pilou97) [reply utils/tzktHooks.ts]: fixed in: https://github.com/marigold-dev/tzsafe-ui/pull/167/commits/d6061725be57076a19783e218f046ffbbed80438

File: utils/tzktHooks.ts (hunk 12 above)
Comment (Pilou97) [reply utils/tzktHooks.ts]: fixé à plusieurs endroits:
 - https://github.com/marigold-dev/tzsafe-ui/pull/167/commits/127303b85a470702c7207fa13c6e54d65a99390c
 - https://github.com/marigold-dev/tzsafe-ui/pull/167/commits/e0d7f92658bf64941315f0fd7f57f3c702792217
 - https://github.com/marigold-dev/tzsafe-ui/pull/167/commits/97ec6759a0b46a923c2fbf948ae9fe2334a5c8bc
 - https://github.com/marigold-dev/tzsafe-ui/pull/167/commits/ba4904b348e74f65794bbb4034c118f30fef55a4
 - https://github.com/marigold-dev/tzsafe-ui/pull/167/commits/df956bff980ef4a2cb1adaa28ef5ff85b6f71443

File: utils/tzktHooks.ts (hunk 10 above)
Comment (Pilou97) [reply utils/tzktHooks.ts]: fixed in: https://github.com/marigold-dev/tzsafe-ui/pull/167/commits/1491ff68e23888cdbc315a329c0c95e66375bf85

File: components/dashboard/index.tsx (hunk 5 above)
Comment (Pilou97) [reply components/dashboard/index.tsx]: fixed in: https://github.com/marigold-dev/tzsafe-ui/pull/167/commits/456e29db1154dddebeeac6f64d6e39a3de00d72d

File: components/dashboard/donut.tsx (hunk 4 above)
Comment (Pilou97) [reply components/dashboard/donut.tsx]: fixed in: https://github.com/marigold-dev/tzsafe-ui/pull/167/commits/60f7cff4040219dce9d2c33309a2b3908e80a38a

File: components/dashboard/donut.tsx (hunk 3 above)
Comment (Pilou97) [reply components/dashboard/donut.tsx]: fixed in:
https://github.com/marigold-dev/tzsafe-ui/pull/167/commits/b25f92837edad6ec30404151f46ae961dc208fe8

File: components/dashboard/balance.tsx (hunk 2 above)
Comment (Pilou97) [reply components/dashboard/balance.tsx]: fixed in: https://github.com/marigold-dev/tzsafe-ui/pull/167/commits/34232e4aa6aed7278eb8921d2554a358cf208fca

File: components/dashboard/assets.tsx (hunk 1 above)
Comment (Pilou97) [reply components/dashboard/assets.tsx]: fixed in: https://github.com/marigold-dev/tzsafe-ui/pull/167/commits/5e90cea124fe094c16f6d29919b28da0fda59b27

File: utils/tzktHooks.ts (hunk 8 above)
Comment (Pilou97) [reply utils/tzktHooks.ts]: fixed: https://github.com/marigold-dev/tzsafe-ui/pull/167/commits/7624e35d0a26f68d2cdc0ca77342b0fc590bbe60

File: utils/tzktHooks.ts (hunk 9 above)
Comment (Pilou97) [reply utils/tzktHooks.ts]: fixed in: https://github.com/marigold-dev/tzsafe-ui/pull/167/commits/7624e35d0a26f68d2cdc0ca77342b0fc590bbe60

File: utils/tzktHooks.ts (hunk 7 above)
Comment (Pilou97) [reply utils/tzktHooks.ts]: fixed in: https://github.com/marigold-dev/tzsafe-ui/pull/167/commits/97ec6759a0b46a923c2fbf948ae9fe2334a5c8bc

**Comment by quentin-burg** [comment]
LGTM 🚀

**Pull request merged**